### PR TITLE
typing(metrics): Various fixes

### PR DESCRIPTION
**Pull request by armenzg**

### CODE REVIEW

File: src/sentry/snuba/metrics_performance.py [hunk 1]
```diff
@@ -100,7 +102,7 @@ def bulk_timeseries_query(
     on_demand_metrics_type: Optional[MetricSpecType] = None,
     groupby: Optional[Column] = None,
     apply_formatting: Optional[bool] = True,
-) -> SnubaTSResult:
+) -> SnubaTSResult | EventsResponse:
```
Comment (armenzg) [inline src/sentry/snuba/metrics_performance.py:105]: When we call with `apply_formatting=False` we return an `EventsResponse`.

File: src/sentry/snuba/metrics_performance.py [hunk 2]
```diff
@@ -113,7 +115,6 @@ def bulk_timeseries_query(
     if metrics_compatible:
         with sentry_sdk.start_span(op="mep", description="TimeseriesMetricQueryBuilder"):
             metrics_queries = []
-            metrics_query = None
```
Comment (armenzg) [inline src/sentry/snuba/metrics_performance.py:116]: The code here is very weird. First, we iterate over `queries`, which means that the last query will be the one setting the value of `metrics_query` which later will have `process_results()` called in line 142.

**Review by evanh — APPROVED** [review]

File: src/sentry/utils/snuba.py [hunk 3]
```diff
@@ -810,7 +810,7 @@ def bulk_snql_query(
     requests: List[Request],
     referrer: Optional[str] = None,
     use_cache: bool = False,
-) -> Mapping[str, Any]:
+) -> List[str, Any]:
```
Comment (armenzg) [inline src/sentry/utils/snuba.py:813]:
```suggestion
) -> dict[str, Any]:
```

File: src/sentry/snuba/discover.py [hunk 4]
```diff
@@ -69,6 +69,9 @@
 
 class EventsMeta(TypedDict):
     fields: Dict[str, str]
+    datasetReason: Optional[str]
+    isMetricsData: Optional[bool]
+    isMetricsExtractedData: Optional[bool]
```
Comment (asottile-sentry) [inline src/sentry/snuba/discover.py:74]: are these `Optional[str]` (aka, either a `str` or `None`) or are these `NotRequired[str]` (either a `str` or it is missing) ?

Comment (armenzg) [reply src/sentry/snuba/discover.py:74]: @asottile-sentry not required. Thanks for catching it! I was not aware of it

**Review by asottile-sentry — APPROVED** [review]

**Review by asottile-sentry — APPROVED** [review]

File: src/sentry/snuba/metrics_performance.py [hunk 5]
```diff
@@ -133,18 +134,18 @@ def bulk_timeseries_query(
 
             metrics_referrer = referrer + ".metrics-enhanced"
             bulk_result = bulk_snql_query(metrics_queries, metrics_referrer)
-            result = {"data": []}
+            _result: dict[str, Any] = {"data": []}
             for br in bulk_result:
-                result["data"] = [*result["data"], *br["data"]]
-                result["meta"] = br["meta"]
+                _result["data"] = [*_result["data"], *br["data"]]
```
[the remark this answers is not in the document]
Comment (armenzg) [reply src/sentry/snuba/metrics_performance.py:139]: For the curious, this is what did the trick:
https://github.com/getsentry/sentry/pull/60614/commits/eb3e46d0a34f1f184a0b238c2017bbcafaf63a34

**Pull request merged**